Refactors Throw Impact

Kurfursten opened the pull request:
Checks to see if a thrown object impacts are now OOP

Most important consequence of this is that shields can block thrown items, fixes #22204

Also refactored on_block to take the thrown item and stop it, but objects that don't inherit base behavior don't use this (e.g.: if you block a toolbox with a katana it will stop on your tile, if you dodge it with the hologram projector it will fly through to the end of its throw course)

This also means that it is much easier to manipulate the blocked item in future PRs. For example, if someone wanted to add a behavior where blocking a thrown item with a katana reflects it back at the thrower, they could do that.

🆑 
* tweak: Shields and other blocking items (e.g.: katanas) now properly block thrown objects

ShiftyRail commented:
Please leave bolas out of this. They are situational enough already.

Kurfursten commented:
This is a significant buff to riot shields, sec main

I do imagine you're just memeing though, since it would be hard to imagine someone seriously arguing shields should block melee attacks, projectiles, bullets, but not thrown objects. The main use riot shields get irl is blocking thrown rocks.

ShiftyRail commented:
Riot shield (and katanas) are absurdly strong already, bolas are a good and natural way to bypass them. It is also available to greyshirts so they can be more of a challenge.

Bolas target the legs, so it's silly that something protecting the chest blocks them. In my opinion.

Kurfursten commented:
Just so we're on the same page here, am I getting this right?

**Reasonable:** An object no wider than a fifth of an inch traveling at 1700 mph blocked with the narrow blade of a katana
**Silly:** A bulky throwing object about the size of three fists traveling likely no faster than 30mph blocked by slightly lowering a shield which is physically wider than one's entire body

DamianX commented:
muh bolas

ShiftyRail commented:
I might be biased, but I think that bolas are criminally underused and that shields are already very (almost too) good at protecting their wielder against ranged attacks. 
It makes sense to me that bolas bypass those defensive checks, as they target the lower part of the body and stuff.

>A 5mm bullet traveling at 2735kmh blocked with the 30mm width of a katana

He's a ninja with god-like reflexes

DamianX commented:
I think there's an issue this closes, something about snowballs.
Also since we're all repeating the things we've already said: I agree with shifty, don't nerf muh bolas

Sooose commented:
The bola boys will remember this.

boy2mantwicethefam commented:
I have pity for your soul for digging through armor code and witnessing what I've seen

bathosbathtime commented:
The dismissive attitude towards the opinions of the players that would most use both sides of this change is bad.
Further, do you know what niche bolas fit in the game?  Why anyone would use them over any other form of disable?  They are effective regardless of the target's equipment.  This argument for realism fails to address the reason why bolas are used at all in game and moves them from a high risk niche item to actual joke item.

jknpj commented:
Security doesn't need to have a counter for everything. This is tabe drama all over again.

ShiftyRail commented:
Bolas fill a niche that is much more immersion-friendly than tabe, is not the exclusive privelege of sec (by far) and with a more interesting gameplay than trapping simple mobs.

jknpj commented:
It's OK to make bolas good against shields, but make them straight up IGNORE shields?

Kurfursten commented:
Bathos, I sincerely feel you are being the dismissive one on this. You think I dug through this particularly painfully structured code and refactored it for fun or to antagonize people? I fixed this intensely buggy and unintuitive behavior *because* of player confusion. The reason I was able to even identify this bug was that people have been complaining to me in ahelp or deadchat that hologram projectors "don't work".

Every object in this game - *thousands of individual things* - can be thrown. Removing this bug fixes all of them, it's not targeted at bolas any more than it is targeted at floor tiling people.

SonixApache commented:
just shitcode an exception for bolas in later

Kurfursten commented:
There's a way you could do it with less shitcode which is to make some kind of shield-ignore-bonus var and you could even give it to other things like hammers or something (in classic 'hammer beats armor' style). If someone wants to refactor the shield percent balance that's not atomic to this though - I'm not touching that. And I know I do meme about unatomic a lot but this is a case where it would matter since I think it would be separately contentious.

codeungabunga commented:
I don't understand how you can claim that bolas are underused when you use them at every opportunity.

SonixApache commented:
bolas are underused in the sense that they're pretty situational (use against targets that are resistant to tasers) and not many people carry/use them because of that

Tokanova commented:
bolas are gay

bathosbathtime commented:
To take someone's argument and tell them they must be memeing infuriates me, but I don't think, now or ever, you made this PR to shitcan bolas.  It is good that this bug is getting fixed but you should at least recognize that it puts bolas in a poor spot even if you don't change the PR to compensate.  To have them have the same chance as a taser bolt to go means the taser can do what the bolas can without giving the target a weapon if you miss.  They can still serve the niche at slowing down laser resistant targets or absolutely replacing shoecuffs but they will definitely be used less which makes me sad as they are an interesting solution to the problem of "I want to stun this person" while having risk.
In any event, dismissing someone so out of hand is disgusting and I will always hate such behavior.  Throw code being bugged is a huge problem though and is clearly more important than either issue I have here, so this is a good PR.

Kurfursten commented:
I assumed he was memeing because right before making that post, he replied to it "muh bolas" in the Discord. With that context it made it seem like he was making a joke. e.g.: "Leave my bolas out of this" ends up sounding like "Leave Brittney alone"

bathosbathtime commented:
> d*scord
>
Ah, well there's your problem.  Still, he goes into a bit of detail in his second comment.  I suppose he could be extremely invested in the joke but that's not a conclusion I would be comfortable with myself.

ShiftyRail commented:
I just said "muh bolas" in d*scord to use my god-given 1st amendment right to shitpost, everything I post on the Git is at least semi-serious
